HW2/P5/physics.pyx

```diff
@@ -0,0 +1,167 @@
+    # objects to check for collisions.
+
+    x_idx = <int>(sqrt(2) * <float>(XY[i, 0]) / R)
+    y_idx = <int>(sqrt(2) * <float>(XY[i, 1]) / R)
+
+    for j in range (x_idx-1, x_idx+2):
+        for k in range (y_idx-1, y_idx+2):
+            index = Grid[j,k]
+            if index == i or index == -1 or index >= count or j < 0 or k < 0 \
+                    or j >= Grid.shape[0] or k >= Grid.shape[1]:
+                continue  # Skip self and check bounds
+            XY2 = &(XY[index, 0])
+            V2 = &(V[index, 0])
+
+            # Always lock lower lock first
+            if i > index :
+                acquire(&(locks[i]))
+                acquire(&(locks[index]))
+            else:
+                acquire(&(locks[index]))
+                acquire(&(locks[i]))
+
+            if overlapping(XY1, XY2, R):
+                # SUBPROBLEM 4: Add locking
+                if not moving_apart(XY1, V1, XY2, V2):
+                    collide(XY1, V1, XY2, V2)
+                # give a slight impulse to help separate them
+                for dim in range(2):
+                    V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+            if i > index:
+                release(&(locks[i]))
+                release(&(locks[index]))
+            else:
+                release(&(locks[index]))
+                release(&(locks[i]))
+
+
+cpdef update(FLOAT[:, ::1] XY,
+             FLOAT[:, ::1] V,
+             UINT[:, ::1] Grid,
+             float R,
+             float grid_spacing,
+             uintptr_t locks_ptr,
+             float t):
+    cdef:
+        int count = XY.shape[0]
+        int i, j, dim
+        FLOAT *XY1, *XY2, *V1, *V2
+        float x, y
+        # SUBPROBLEM 4: uncomment this code.
+        omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
+
+    assert XY.shape[0] == V.shape[0]
+    assert XY.shape[1] == V.shape[1] == 2
+
+    with nogil:
+        # bounce off of walls
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in prange(count, num_threads=4, schedule="static", chunksize=count/4):
+            for dim in range(2):
+                if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
+                    ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
+                    V[i, dim] *= -1
+
+        # bounce off of each other
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in prange(count, num_threads=4, schedule="static", chunksize=count/4):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, locks)
+
+        # update positions
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
+        #    scheduling).
+        # SUBPROBLEM 2: update the grid values.
+        for i in prange(count, num_threads=4, schedule="static", chunksize=count/4):
+            x = 0
+            y = 0
+            for dim in range(2):
```

Comment:
It is a little confusing to have a loop over the dimensions and then condition on each one, I would either get rid of the condition and add the updated values to a array of length dim or just update x and y separately.